include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,275 @@
+
+			auto checkFormat = state->inImage->getCreationParameters().format;
+			if (isIntegerFormat(checkFormat))
+				return executeInterprated(state, reinterpret_cast<uint64_t*>(state->scratchMemory));
+			else
+				return executeInterprated(state, reinterpret_cast<double*>(state->scratchMemory));
+		}	
+
+	private:
+
+		template<typename decodeType> //!< double or uint64_t
+		static inline bool executeInterprated(state_type* state, decodeType* scratchMemory)
+		{
+			const asset::E_FORMAT inFormat = state->inImage->getCreationParameters().format;
+			const auto inTexelByteSize = asset::getTexelOrBlockBytesize(inFormat);
+			const auto currentChannelCount = asset::getFormatChannelCount(inFormat);
+			const auto arrayLayers = state->inImage->getCreationParameters().arrayLayers;
+			static constexpr auto maxChannels = 4u;
+
+			#ifdef _NBL_DEBUG
+			memset(scratchMemory, 0, state->scratchMemoryByteSize);
+			#endif // _NBL_DEBUG
+
+			const core::vector3du32_SIMD scratchByteStrides = TexelBlockInfo(asset::E_FORMAT::EF_R64G64B64A64_SFLOAT).convert3DTexelStridesTo1DByteStrides(state->extentLayerCount);
+			const auto scratchTexelByteSize = scratchByteStrides[0];
+
+			// I wonder if we should let somebody pass through more than 1 layer, though I find it cool
+
+			const auto&& [copyInBaseLayer, copyOutBaseLayer, copyLayerCount] = std::make_tuple(state->inBaseLayer, state->outBaseLayer, state->layerCount);
+			state->layerCount = 1u;
+			 
+			auto resetState = [&]()
+			{
+				state->inBaseLayer = copyInBaseLayer;
+				state->outBaseLayer = copyOutBaseLayer;
+				state->layerCount = copyLayerCount;
+			};
+
+			for (uint16_t w = 0u; w < copyLayerCount; ++w)
+			{
+				std::array<decodeType, maxChannels> maxAbsoluteDecodeValues = {};
+
+				{
+					const uint8_t* inData = reinterpret_cast<const uint8_t*>(state->inImage->getBuffer()->getPointer());
+					const auto blockDims = asset::getBlockDimensions(state->inImage->getCreationParameters().format);
+					static constexpr uint8_t maxPlanes = 4;
+
+					auto decode = [&](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+					{
+						core::vectorSIMDu32 localOutPos = readBlockPos * blockDims - core::vectorSIMDu32(state->inOffset.x, state->inOffset.y, state->inOffset.z);
+
+						auto* inDataAdress = inData + readBlockArrayOffset;
+						const void* inSourcePixels[maxPlanes] = { inDataAdress, nullptr, nullptr, nullptr };
+
+						decodeType decodeBuffer[maxChannels] = {};
+						for (auto blockY = 0u; blockY < blockDims.y; blockY++)
+							for (auto blockX = 0u; blockX < blockDims.x; blockX++)
+							{
+								asset::decodePixelsRuntime(inFormat, inSourcePixels, decodeBuffer, blockX, blockY);
+								const size_t offset = asset::IImage::SBufferCopy::getLocalByteOffset(core::vector3du32_SIMD(localOutPos.x + blockX, localOutPos.y + blockY, localOutPos.z), scratchByteStrides);
+								memcpy(reinterpret_cast<uint8_t*>(scratchMemory) + offset, decodeBuffer, scratchTexelByteSize);
```

Comment:
because the decode buffer being tightly packed (stride=width) this copy should look a bit different

Reply:
also move the divide by `z` here

Reply:
we can't move it there because of overlapping, but we can ensure flatten gets executed previously and then perform the division in the lambda

Reply:
I think it could break it all otherwise

Reply:
okay I realised it won't be a problem, TODO

Reply:
corrected in following commit